include/nbl/builtin/hlsl/blit/common.hlsl

```diff
@@ -0,0 +1,147 @@
+// Copyright (C) 2023 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_BLIT_COMMON_INCLUDED_
+#define _NBL_BUILTIN_HLSL_BLIT_COMMON_INCLUDED_
+
+#include <nbl/builtin/hlsl/cpp_compat/cpp_compat.hlsl>
+#include <nbl/builtin/hlsl/blit/formats_encode.hlsl>
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace blit
+{
+
+template <uint32_t AlphaBinCount>
+struct AlphaStatistics
+{
+	uint32_t passedPixelCount;
+	uint32_t histogram [AlphaBinCount];
+};
+
+template <uint32_t Dimension>
+struct dim_to_in_texture { using type = void; };
+
+template <>
+struct dim_to_in_texture<1> { using type = Texture1DArray<float4>; };
+
+template <>
+struct dim_to_in_texture<2> { using type = Texture2DArray<float4>; };
+
+template <>
+struct dim_to_in_texture<3> { using type = Texture3D<float4>; };
+
+
+template<uint32_t Dimension, bool RequiresSoftwareEncode>
+struct dim_and_software_encode_to_out_texture {};
+
+// These have to be specialized for both dimension and software encode
+// instead of defining seperate trait for the element type because
+// for some reason DXC complains when using inferred types for RWTexture
+template <>
+struct dim_and_software_encode_to_out_texture<1, true>
+{
+	using type = RWTexture1DArray<uint32_t4>;
+};
+
+template <>
+struct dim_and_software_encode_to_out_texture<2, true>
+{
+	using type = RWTexture2DArray<uint32_t4>;
+};
+
+template <>
+struct dim_and_software_encode_to_out_texture<3, true>
+{
+	using type = RWTexture3D<uint32_t4>;
+};
+
+template <>
+struct dim_and_software_encode_to_out_texture<1, false>
+{
+	using type = RWTexture1DArray<float4>;
+};
+
+template <>
+struct dim_and_software_encode_to_out_texture<2, false>
+{
+	using type = RWTexture2DArray<float4>;
+};
+
+
+template <>
+struct dim_and_software_encode_to_out_texture<3, false>
+{
+	using type = RWTexture3D<float4>;
+};
```

Comment:
thanks to shaderStorageImageExtendedFormats you only need to care about `float4`, so you can use a simpler way to do it just as you did with `dim_to_texture`